### PR TITLE
Add `shadow-style` attribute - #28

### LINKED ISSUE #28: Custom styles
I would like to be able to set styles on the editor, but I'm having a hard time doing it. Could this component possibly provide some custom CSS properties for styling?

### PR DESCRIPTION
**Pull request by tomalec**
Add `shadow-style` - new attribute to declaratively inject styles to shadow DOM,
add test case for custom Ace themes

fixes #28

### CODE REVIEW

File: juicy-ace-editor.html [hunk 1]
```diff
@@ -158,6 +160,12 @@
                 case "wrapmode":
                     this.editor.getSession().setUseWrapMode( newVal !== null );
                     break;
+                // non-Ace specific
+                case 'shadow-style':
+                    if(oldVal){
+                        this.shadowRoot.querySelector('#'+oldVal).remove();
```
Comment (alshakero) [inline juicy-ace-editor.html]: I think this `#` isn't needed, everywhere else it's documented as a CSS selector while this assumes an ID.

File: juicy-ace-editor.html [hunk 2]
```diff
@@ -169,8 +177,12 @@
...
         }
-        injectTheme(themeId){
-            var n = document.querySelector(themeId);
+        /**
+         * Injects a style element into juicy-ace-editor's shadow root
+         * @param {CSSSelector} selector for an element in the same shadow tree or document as `juicy-ace-editor`
+         */
+        injectTheme(selector){
+            var n = this.getRootNode().querySelector(selector);
             this.shadowRoot.appendChild(cloneStyle(n));
```
Comment (alshakero) [inline juicy-ace-editor.html]: I checked `cloneStyle` function, not sure why not use `cloneNode` directly or inside the helper function?

File: test/shadow-style.html [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    <!-- or v1 -->
+    <script src="../../webcomponentsjs/webcomponents-lite.js"></script>
+    <script src="../../document-register-element/build/document-register-element.js"></script>
+    <script src="../../web-component-tester/browser.js"></script>
+
+    <link rel="import" href="../juicy-ace-editor.html">
+</head>
+
+<body>
+    <test-fixture id="hadmadeStyle">
```
Comment (alshakero) [inline test/shadow-style.html]: Hand?

File: juicy-ace-editor.html [hunk 4]
```diff
@@ -113,6 +113,8 @@
                 session.setUseSoftTabs( this.getAttribute("softtabs") );
                 this.getAttribute("tabsize") && session.setTabSize( this.getAttribute("tabsize") );
                 session.setUseWrapMode( this.hasAttribute("wrapmode") );
+                // non ace specific
+                this.hasAttribute("shadow-style") && this.injectTheme(this.getAttribute('shadow-style'));
```
Comment (alshakero) [inline juicy-ace-editor.html]: I suggest adding support for `shadowStyle` prop too. If you agree I think it should support `HTMLStyleElement`s and `string` CSS Selectors. This gives so much more freedom for the developer.

File: test/shadow-style.html [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            // wait for theme loaded
+            setTimeout(done, 100);
+        });
+
+        it('should stamp it to its shadowRoot and apply styles', function() {
+            expect(juicyElement.shadowRoot.querySelector('style#my-theme-id')).to.be.not.null;
+            expect(window.getComputedStyle(juicyElement.shadowRoot.querySelector('.ace_gutter')).fontSize).to.be.equal('7px');
+        });
+    });
+</script>
```
Comment (alshakero) [inline test/shadow-style.html]: If you agree to support adding a property that supports adding `HTMLStyleElement`s as they're (as opposed to selectors). Please add another test for it here.

```js
const style = document.querySelector('#my-style');
juicyElement.style = style;
expect(...);
.
.
.
```

File: juicy-ace-editor.html (hunk 4 above)
Comment (tomalec) [reply juicy-ace-editor.html]: We don't support any other properties than `value` and for YAGNI, I'd leave it that way, unless somebody with real use case asks for it.

**Comment by tomalec** [comment]
I covered most of your comments, but I'd leave `shadowStyles` feature for future, as we do not support any other properties (except `value`). WDYT?

**Comment by tomalec** [comment]
I agree with ergonomics of properties and object refs, but let's do it once we need it. And move with this one sooner.

**Pull request merged**